### PR TITLE
return error codes to application

### PR DESCRIPTION
**Pull request by chrimaka**
That's a PR for passing the error codes from the stack to the application. It requires a bit more fine tuning (null checks; casting of CBErrors from iOS). But this would be the general concept. Feedback very welcome.

### CODE REVIEW

**Comment by janusw** [comment]
Thanks for the PR! Unfortunately the build fails with errors like this:

```
build: Source/Plugin.BLE/Windows/Characteristic.cs#L50
'Characteristic.ReadNativeAsync()': return type must be 'Task<Tuple<byte[], int>>' to match overridden member 'CharacteristicBase<GattCharacteristic>.ReadNativeAsync()'
```

Please fix! :)

**Comment by janusw** [comment]
> @janusw I have a question on the build process. I was able to build the former 2.1.x "out of the box". I also can build the `Plugin.BLE.sln` without issues, but If I try to build `BLE.sln`, VS2022 for macOS (17.5.3 (build 15)) already gives me an error during opening ` error NETSDK1045: Das aktuelle .NET SDK unterstützt .NET Core 6.0 nicht als Ziel. Geben Sie entweder .NET Core 5.0 oder niedriger als Ziel an, oder verwenden Sie eine .NET SDK-Version, die .NET Core 6.0 unterstützt. Für das Projekt kann keine Paketspezifikation erstellt werden.`

I'm afraid that's some sort of deficiency of VS4Mac. VS on Windows should work AFAIK. Maybe we can find a workaround ...

File: Source/Plugin.BLE/Android/GattCallback.cs [hunk 1]
```diff
@@ -174,7 +174,7 @@ public override void OnCharacteristicRead(BluetoothGatt gatt, BluetoothGattChara
 
             Trace.Message("OnCharacteristicRead: value {0}; status {1}", characteristic.GetValue().ToHexString(), status);
 
-            CharacteristicValueRead?.Invoke(this, new CharacteristicReadCallbackEventArgs(characteristic));
+            CharacteristicValueUpdated?.Invoke(this, new CharacteristicReadCallbackEventArgs(characteristic, status));
```
Comment (janusw) [inline Source/Plugin.BLE/Android/GattCallback.cs:177]: Why do you change the invoked event to `CharacteristicValueUpdated` here? This leads to the warning:
GattCallback.cs(72,72): Warning CS0067: The event 'GattCallback.CharacteristicValueRead' is never used (CS0067)

**Review by janusw — COMMENTED** [review]
Overall this looks pretty good to me, but I have some minor comments ...

**Comment by janusw** [comment]
> @janusw I have a question on the build process. I was able to build the former 2.1.x "out of the box". I also can build the `Plugin.BLE.sln` without issues, but If I try to build `BLE.sln`, VS2022 for macOS (17.5.3 (build 15)) already gives me an error during opening

This is fixed (or rather worked around) by now via #685. Using `BLE.Mac.slnf` should work on Mac (however, it only includes the core libraries, but not all the sample clients).

File: Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs [hunk 2]
```diff
@@ -79,25 +79,25 @@ public interface ICharacteristic
         /// Reads the characteristic value from the device. The result is also stored inisde the Value property.
         /// </summary>
         /// <param name="cancellationToken"></param>
-        /// <returns>A task that represents the asynchronous read operation. The Result property will contain the read bytes.</returns>
+        /// <returns>A task that represents the asynchronous read operation. The Result property will contain a tuple with the read bytes and the ble status code.</returns>
         /// <exception cref="InvalidOperationException">Thrown if characteristic doesn't support read. See: <see cref="CanRead"/></exception>
         /// <exception cref="CharacteristicReadException">Thrown if the reading of the value failed.</exception>
-        Task<byte[]> ReadAsync(CancellationToken cancellationToken = default);
+        Task<Tuple<byte[], int>> ReadAsync(CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (janusw) [reply Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs:85]: Note that this comment is kind of optional. I will not force you to change it, I just wanted to mention it. If you have considered other options, but prefer the current solutions, I'm fine with that.

**Comment by janusw** [comment]
@chrimaka Could you take care of the remaining comments, or should I fix them (so that we can move forward with this)? It would be nice if we could merge it soon.

**Comment by chrimaka** [comment]
@janusw sorry for the late reply. I was out sick for some time. I'll look into this on Thrusday/Friday + Weekend.

File: Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrimaka) [reply Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs:85]: I changed this to native c# net7 Tuple which are named. Way better. Thanks for pointing that out.

File: Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs [hunk 3]
```diff
@@ -79,25 +79,25 @@ public interface ICharacteristic
...
 
         /// <summary>
         /// Sends <paramref name="data"/> as characteristic value to the device.
         /// </summary>
         /// <param name="data">Data that should be written.</param>
         /// <param name="cancellationToken"></param>
         /// <returns>
-        /// A task that represents the asynchronous read operation. The Task will finish after the value was written. The Result property will be <c>true</c> if the value
-        /// was written successful, otherwise <c>false</c>.
+        /// A task that represents the asynchronous read operation. The Task will finish after the value was written. The Result property will the errror
```
[the remark this answers is not in the document]
Comment (chrimaka) [reply Source/Plugin.BLE/Shared/Contracts/ICharacteristic.cs:93]: Fixed. Thanks for the note.

File: Source/Plugin.BLE/Android/GattCallback.cs (hunk 1 above)
Comment (chrimaka) [reply Source/Plugin.BLE/Android/GattCallback.cs:177]: you are right. Copy and paste error. My bad.

**Review by janusw — APPROVED** [review]
Awesome, thanks for addressing all my comments. This is ok for master by now.

**Pull request merged**